File: src/topology/show_switch_status.c
/*
 * Copyright (C) 2012 NEC Corporation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */


#include <assert.h>
#include <getopt.h>
#include <stdio.h>
#include <unistd.h>
#include "trema.h"
#include "libtopology.h"
#include "show_topology.h"
#include "topology_service_interface_option_parser.h"


void
usage() {
  topology_service_interface_usage( get_executable_name(), "Show switch status", NULL );
}


static void
timed_out( void *user_data ) {
  UNUSED( user_data );

  error( "timed out." );

  stop_trema();
}


void
print_port_status( void *user_data, size_t number, const topology_port_status *ports ) {
  UNUSED( user_data );
  const topology_port_status *port = ports;
  
  printf( "Port status\n" );
  for ( size_t i = 0; i < number; i++, port++ ) {
    const char *status = "unknown";
    switch ( port->status ) {
      case TD_PORT_DOWN:
        status = "down";
        break;
      case TD_PORT_UP:
        status = "up";
        break;
    }
    printf( "  dpid : 0x%" PRIx64 ", port : 0x%d(%s), status : %s\n", 
            port->dpid, port->port_no, port->name, status );
  }
  
  stop_trema();
}


void
print_switch_status( void *user_data, size_t number, const topology_switch_status *switches ) {
  UNUSED( user_data );
  const topology_switch_status *sw = switches;

  printf( "Switch status\n" );
  for ( size_t i = 0; i < number; i++, sw++ ) {
    const char *status = "unknown";
    switch ( sw->status ) {
      case TD_SWITCH_DOWN:
        status = "down";
        break;
      case TD_SWITCH_UP:
        status = "up";
        break;
    }
    printf( "  dpid : 0x%" PRIx64 ", status : %s\n", sw->dpid, status );
  }
  
  get_all_port_status( print_port_status, NULL );
}


int
main( int argc, char *argv[] ) {
  init_trema( &argc, &argv );
  init_topology_service_interface_options( &argc, &argv );
  init_libtopology( get_topology_service_interface_name() );

  get_all_switch_status( print_switch_status, NULL );

  add_periodic_event_callback( 15, timed_out, NULL );

  start_trema();

  finalize_libtopology();
  finalize_topology_service_interface_options();

  return 0;
}


/*
 * Local variables:
 * c-basic-offset: 2
 * indent-tabs-mode: nil
 * End:
 */

File: src/examples/topology/show-switch-status.rb
#
# Copyright (C) 2008-2013 NEC Corporation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License, version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#


class ShowSwitchStatus < Controller
  include Topology


  oneshot_timer_event :timed_out, 15
  oneshot_timer_event :on_start, 0
  def on_start
    get_all_switch_status
  end

  def all_switch_status sw_status
    puts "Switch status"
    sw_status.each do | sw_hash |
      sw = Topology::Switch.new( sw_hash )

      status_str = "unknown"
      if sw.up? then
        status_str = "up"
      else
        status_str = "down"
      end
      puts "  dpid : 0x#{sw.dpid.to_s(16)}, status : #{status_str}"
    end

    get_all_port_status
  end

  def all_port_status port_status
    puts "Port status"
    port_status.each do | port_hash |
      port = Port.new( port_hash )
      status_str = "unknown"
      if port.up? then
        status_str = "up"
      else
        status_str = "down"
      end
      
      puts "  dpid : 0x#{port.dpid.to_s(16)}, port : #{port.portno.to_s}(#{port.name}), status : #{status_str}, external : #{port.external?.to_s}"
    end

    shutdown!
  end


  def timed_out
    error "timed out."
    shutdown!
  end
end
